File: archive/r/rust/job-sequencing.rs
use std::env::args;
use std::process::exit;
use std::str::FromStr;
use std::cmp::Ordering;

fn usage() -> ! {
    println!("Usage: please provide a list of profits and a list of deadlines");
    exit(0);
}

fn parse_int<T: FromStr>(s: &str) -> Result<T, <T as FromStr>::Err> {
    s.trim().parse::<T>()
}

fn parse_int_list<T: FromStr>(s: &str) -> Result<Vec<T>, <T as FromStr>::Err> {
    s.split(',')
        .map(parse_int)
        .collect::<Result<Vec<T>, <T as FromStr>::Err>>()
}

#[derive(Debug, Ord, Eq)]
struct JobInfo {
    job_id: usize,
    profit: i32,
    deadline: usize,
}

impl JobInfo {
    fn new(job_id: usize, profit: i32, deadline: usize) -> Self {
        Self {job_id: job_id, profit: profit, deadline: deadline}
    }
}

impl PartialOrd for JobInfo {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        // Reverse order of compare so that it is in descending order by profit
        // then deadline
        match self.profit != other.profit {
            true => Some(other.profit.cmp(&self.profit)),
            false => Some(other.deadline.cmp(&self.deadline)),
        }
    }
}

impl PartialEq for JobInfo {
    fn eq(&self, other: &Self) -> bool {
        self.profit == other.profit && self.deadline == other.deadline
    }
}

// Job sequencing with deadlines
// Source: https://www.techiedelight.com/job-sequencing-problem-deadlines/
fn job_sequencing(profits: &Vec<i32>, deadlines: &Vec<i32>) -> Vec<JobInfo> {
    // Set up job details
    let mut jobs: Vec<JobInfo> = profits.iter()
        .zip(deadlines.iter())
        .enumerate()
        .map(|(n, (&p, &d))| JobInfo::new(n + 1, p, d as usize))
        .collect();

    // Get longest deadline
    let longest_deadline: i32 = *deadlines.iter()
        .max()
        .unwrap();

    // Initialize job slots
    let mut slots: Vec<JobInfo> = (0..longest_deadline)
        .map(|_| JobInfo::new(0, 0, 0))
        .collect();

    // Sort jobs by profit then deadline
    jobs.sort();

    // For each job, see if there is available slot at or before the deadline.
    // If so, store this job in that slot
    for job in jobs {
        for j in (0..job.deadline).rev() {
            if slots[j].job_id < 1 {
                slots[j] = job;
                break;
            }
        }
    }

    slots
}

fn get_total_profit(jobs: &Vec<JobInfo>) -> i32 {
    jobs.iter()
        .map(|x| x.profit)
        .sum()
}

fn main() {
    let mut args = args().skip(1);

    // Convert 1st command-line argument to list of integers
    let mut profits: Vec<i32> = args
        .next()
        .and_then(|s| parse_int_list(&s).ok())
        .unwrap_or_else(|| usage());

    // Convert 2nd command-line argument to list of integers
    let mut deadlines: Vec<i32> = args
        .next()
        .and_then(|s| parse_int_list(&s).ok())
        .unwrap_or_else(|| usage());

    // Exit if profits not same length as deadlines
    if profits.len() != deadlines.len() {
        usage();
    }

    // Get job sequence
    let jobs = job_sequencing(&profits, &deadlines);

    // Get total profit and display
    println!("{}", get_total_profit(&jobs));
}

File: archive/c/c/job-sequencing.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_JOBS 100

typedef struct {
    int profit;
    int deadline;
} Job;

int compare(const void* a, const void* b) {
    return ((Job*)b)->profit - ((Job*)a)->profit;
}

int min(int a, int b) {
    return (a < b) ? a : b;
}

int max(int a, int b) {
    return (a > b) ? a : b;
}

int jobSequencing(Job jobs[], int n) {
    qsort(jobs, n, sizeof(Job), compare);

    int maxDeadline = 0;
    for (int i = 0; i < n; i++) {
        maxDeadline = max(maxDeadline, jobs[i].deadline);
    }

    int slot[MAX_JOBS] = {0};
    int totalProfit = 0;

    for (int i = 0; i < n; i++) {
        for (int j = min(n, jobs[i].deadline) - 1; j >= 0; j--) {
            if (slot[j] == 0) {
                slot[j] = 1;
                totalProfit += jobs[i].profit;
                break;
            }
        }
    }

    return totalProfit;
}

int* parseInput(char* input, int* size) {
    int* arr = malloc(MAX_JOBS * sizeof(int));
    char* token = strtok(input, ", ");
    *size = 0;
    while (token != NULL) {
        arr[(*size)++] = atoi(token);
        token = strtok(NULL, ", ");
    }
    return arr;
}

int main(int argc, char* argv[]) {
    if (argc != 3) {
        printf("Usage: please provide a list of profits and a list of deadlines\n");
        return 1;
    }

    int profitSize, deadlineSize;
    int* profits = parseInput(argv[1], &profitSize);
    int* deadlines = parseInput(argv[2], &deadlineSize);

    if (profitSize != deadlineSize) {
        printf("Usage: please provide a list of profits and a list of deadlines\n");
        free(profits);
        free(deadlines);
        return 1;
    }

    Job jobs[MAX_JOBS];
    for (int i = 0; i < profitSize; i++) {
        jobs[i].profit = profits[i];
        jobs[i].deadline = deadlines[i];
    }

    int result = jobSequencing(jobs, profitSize);
    printf("%d\n", result);

    free(profits);
    free(deadlines);
    return 0;
}
